New feature: panel resizing with mouse

yulian5 opened the pull request:
<!-- Thank you for contributing! Please follow the steps below -->

<!-- Enter a brief description of your PR here -->
## Summary
The PR introduces a new feature to Far Manager: panel resizing with mouse.

<!-- If this PR is relevant to any other issues or existing PRs, link them here --> 
## References

<!-- Please review the items on the PR checklist before submitting -->
## Checklist
- [ X] I have followed the [contributing guidelines](https://github.com/FarGroup/FarManager/blob/master/CONTRIBUTING.md).
- [ ] I have discussed this with project maintainers: <!-- add a link to the corresponding issue / discussion / forum topic here --> <br/>
If not checked, I accept that this work might be rejected in favor of a different grand plan.<br/>

<!-- Enter a more detailed description of the PR and any additional comments here -->
## Details
This feature introduces the ability to resize file panels in Far Manager using the mouse.

Users can now:

- Hover over the border between the left and right panels to see visual feedback indicating a resizable area (grip).
- Click and drag the border left or right to dynamically adjust the width of the panels.
- Double-click on the border to quickly restore the default 50/50 equal width split for the panels.

This enhancement provides a more intuitive and efficient way to manage panel layouts, improving user experience for those who prefer mouse-driven interactions.

It includes updates to:

FarColorW.pas and farcolor.hpp: Defines a new palette color COL_PANELDRAGBORDER for the visual feedback during resizing.

farlang.templ.m4: Adds new localization strings for the "Dragging border" text.

filepanels.cpp and filepanels.hpp: Implements the core logic for mouse event handling, border detection, resizing, and drawing the visual feedback. This includes new MouseState enum, IsMouseOverPanelBorder, DrawBorderFeedback, ClearBorderFeedback, and ResetAllMouseStates functions.

palette.cpp and setcolor.cpp: Initializes and integrates the new COL_PANELDRAGBORDER color into the color palette and settings.

*.hlf.m4 files (e.g., FarEng.hlf.m4, FarRus.hlf.m4): Updates help files for various languages to document the new mouse resizing functionality.

HamRusTal commented:
> Do we really want to introduce a new color for this? It would be a breaking change for all custom themes.

New color would be nice and flexible. But the code has to handle the missing color **gracefully** and use a sensible default replacement from the colors defined by the current theme.

alabuzhev commented:
> Do I need to modify changelog and vbuild.m4

Don't worry about these for now. There might be more changes requested which might take a few days and resolving conflicts every time someone merges into master doesn't make much sense.
Once we're ready to merge I'll let you know and those files can be updated then.

> use a sensible default replacement from the colors defined by the current theme

We do not have such a concept. Palette colors are independent and any missing ones will stay default aqua-on-blue-ish.
It might be useful to discuss and explore this more.
For example, background color in quite a few palette items (e.g. panel borders or column titles) is redundant and can be transparent and inherited from panel text or whatever would be previous in Z-order.
However, it might be too much for this particular PR.
Feel free to create an issue or a discussion.

yulian5 commented:
@yegor-mialyk 
> Do we really want to introduce a new color for this? It would be a breaking change for all custom themes.

@alabuzhev 
> It might be useful to discuss and explore this more.

No, it doesn't break anything. All custom color themes work just fine. Yes, we need a new color. There is enum for it and menu entry. Here is sample, you can compile code and try for yourself.

<img width="648" height="328" alt="color_border" src="https://github.com/user-attachments/assets/8ba39a46-ae09-4ae1-8266-ce8ad0824af2" />

I realized from the beginning that custom color themes obviously do not have a new color. But they can define it now. The fallback scheme is following: if color for dragging border is not defined in the theme then we use foreground color from dragging text and background color from normal text (panel background). And all color themes I checked have both colors defined. And in case they don't have them it will be yellow on blue that can be changed easily (but it should be something really broken):

<img width="648" height="328" alt="custom" src="https://github.com/user-attachments/assets/21d24458-8726-47c9-a09c-d36bce0cc93b" />

It looks natural and work on all themes included in Far by default. It still may be some problems here but I'm sure it's possible to fix.

yulian5 commented:
@HamRusTal 

> New color would be nice and flexible. But the code has to handle the missing color gracefully and use a sensible default replacement from the colors defined by the current theme.

Yes, I explained that. Colors were my first concern in this feature!

yulian5 commented:
@HamRusTal 

> On the screenshot, I see that the {left panel's cell below the scroll bar where the “status” area separator joins the right-hand-side frame} is not highlighted yellow whereas the corresponding right panel's cell on its left is (no scrollbar there though). Is it for a purpose or just a bug?

Yes, I notice that. The scrollbar at the bottom is actually expands more than it depicted and this area is clickable and it scrolls the context down. This looks like the problem with scrollbar implementation. We may need to create new issue for that later.

yulian5 commented:
@HamRusTal 

> BTW Do we need to highlight both panels' frames? IMHO just the one being dragged is enough (at least while dragging). We might still highlight both frames on hover though to indicate that both will be affected.

Actually I had the similar concern. I tried to make single lines while dragging but it looked awful. You can change dragging border color to normal color, it works very nice except you cannot see when it hovered (but if you know then it's fine). It's very possible to make separate colors for havering and dragging, not a problem at all. Do we need it? It maybe very confusing.

The single panel can be resized with mouse so to simplify the logic both panel borders are highlighted.

> Also, it would be nice to handle the Esc key press (or RMB click) while dragging to cancel the resize and restore the initial geometry (similar to Windows' window move/resize except for RMB click).

Yes, it's a very good idea, but that would be a completely separate task. Mouse and key handling in Far is quite complicated even without plugins and macros. Double click works fine for now.

yulian5 commented:
@HamRusTal 

> Actually, the entire panel frame below the scrollbar (and above it too) performs scrolling when clicked.

Yes, I already figured is out. I agreed with you and took one cell from the scrolling area:

<img width="400" height="402" alt="scroll_area" src="https://github.com/user-attachments/assets/cd7f48ef-d3de-4c7f-ae5a-27c7b190fc8c" />


It displayed as grip area but not working yet.

> I am rather suggesting to highlight only one panel's (double-line) border. This will be consistent with resizing of a single visible panel too.

But which one I have to highlight? If I highlight right border and the right panel is hidden? I need to change left/right highlighting depending which panel is hidden And another question: the grip area is actually 2 borders, displaying only have of grip area will be confusing. If it's so important maybe just make it configurable. 

> Double-click does not address the scenario when there's some non-default panel width already and the user starts dragging the panel border (maybe by a misclick). Unfortunately, double-click will not restore the pre-drag non-default layout.

I agree. Let me check that. If it's easy to implement I'll add it otherwise it will another PR.

BTW, I added height resizing with mouse:

https://github.com/user-attachments/assets/7687dabc-e249-4b23-befc-93dae8961306

yulian5 commented:
Hi @alabuzhev 

I uploaded new code which fixed calculation of bottom border around text area for visual feedback during height resizing and fixed SonarQubeCloud issues.

However CodeFactor marked the code with red cross because of complexity GetInputRecordImpl() in keboard.cpp file. That's not my code and I don't think that I need to fix it now. I just added couple small functions to block all keys during resizing and allow only Esc key to revert resizing. Please help with that!

Thanks!

yulian5 commented on far/filelist.cpp:
```diff
@@ -722,7 +722,6 @@ class list_less
 
 		return false;
 	}
-
```
There are still a lot mis-formatted places talking only about filelist.cpp.

yulian5 commented:
I merged all my commits into one and updated changelog and vbuild.m4 to the latest. But GitHub doesn't merge it in PR automatically.

alabuzhev commented on far/FarEng.hlf.m4:
```diff
@@ -1160,6 +1163,30 @@ if the initial panel type differs it will be automatically set to the file panel
+ - ^<wrap>Double-click on the bottom border to restore full panel height.
+ - ^<wrap>Press #Esc# during resizing to cancel and revert to the original panel heights.
+
+ #Note for ConEmu users:#
```
This belongs in ConEmu's documentation, not here.

yulian5 replied:
Yes, you're right, let me remove it.

alabuzhev commented on far/FarEng.hlf.m4:
```diff
@@ -365,9 +365,12 @@ $ #Panel control commands#
+ (only when the command line is empty)
+ Change panels width with mouse          #Drag border between panels#
  Restore default panels width                          #Ctrl+Numpad5#
+                        #Mouse double click on border between panels#
```
Formatting looks strange

yulian5 commented on far/FarEng.hlf.m4:
```diff
@@ -884,8 +887,8 @@ default button or does another similar thing).
  Clicking the #middle mouse button# in the ~panels~@PanelCmd@ has the same
-effect as pressing the #Enter# key with the same modifiers (#Ctrl#, #Alt#,
-#Shift#). If the ~command line~@CmdLineCmd@ is not empty, its contents will be executed.
+ effect as pressing the #Enter# key with the same modifiers (#Ctrl#, #Alt#,
```
@alabuzhev, It took a lot of copy/pasting and editing back and forth, let me fix the remaining issues.

yulian5 commented on far/FarEng.hlf.m4:
```diff
@@ -365,9 +365,12 @@ $ #Panel control commands#
  Change current panel height          #Ctrl+Shift+Up,Ctrl+Shift+Down#
- Change panels width                           #Ctrl+Left,Ctrl+Right#
- (when the command line is empty)
+ Change panels height with mouse       #Drag bottom border of panels#
```
Hi Alex, thank you for the feedback!

I don't quite understand what you mean. But I've improved the appearance of the help section, and by the way, I forgot to add a description of how to restore the height using the mouse. Here's what it looks like now:

<img width="457" height="437" alt="help" src="https://github.com/user-attachments/assets/6a2499a6-5a60-459c-8514-92bf5555b1ca" />

Is it OK? If so, I need to update the information in other languages ​​as well.

alabuzhev replied:
Sorry, github line wraps made it confusing.
Pls ignore this particular comment.

HamRusTal commented:
> < 100 bytes of memory per FilePanels instance

I assume you mean the additional data, yet the added instructions code will be larger I believe (still not an issue IMHO).

yulian5 commented on far/filepanels.cpp:
```diff
@@ -1171,18 +1222,629 @@ bool FilePanels::ProcessMouse(const MOUSE_EVENT_RECORD *MouseEvent)
+	const auto DefaultDragBorderColor = F_YELLOW | B_BLUE;
+
+	// If the current color matches the default palette value, or if it seems uninitialized,
+	// apply the fallback logic (theme doesn't define this color or user hasn't customized it)
```
Hi @alabuzhev,

I realized that we had a significant misunderstanding on this matter: you are talking about custom color themes, while I am talking about the color themes that come bundled with Far. We were trying to explain completely different things to each other.

> If the user hasn't customized it then that's exactly what the user wanted.

Absolutely agree. We cannot support user defined color themes. After introducing any feature with additional colors users may adjust colors to their needs.

> And what if the user wants it to be yellow-on-blue in otherwise red-on-green theme?

You're right, that's logic is broken now for default dragging color. Original "non-optimized" code was 3 times bigger.

The reason for this code was simple: obviously Far cannot be released with bundled **broken** color themes. And I didn't want to touch somebody's code for ethical and possible copyright reasons. I probably should ask in advance. Instead I created function that works as memory patch to compensate missing hovering/dragging color. It works for bundles and user defined color schemes as well. 

For user defined schemes that function doesn't make much sense: after the user changes dragging border color, the code uses that new color, so in this case my code works as a one-time function. 

As can I see the "extra\Addons\Colors" code is under the same license as other code so the bundled color scheme may have one new color. It will be much better that having an emulated color logic (for one-time use in some cases). 

Such kind of logic should be used mostly during the first initialization. As you said we may add something like this separately.

The same algorithm for hovering/dragging color (foreground color from `Panel.DragText` foreground and background color from `Panel.Text` background) works fine for all themes, except:

```
black_and_white.farconfig
Colors_from_admin_essp_ru.farconfig
FARColors242.farconfig
```
Let me figure out the consistent colors for those schemes and I'll submit the updated code.

All other issues are fixed, this one is the last (I hope). Also some language help files were not updated for panel height resizing info, I'm fixing them too.

yulian5 replied:
It's not clear what do you mean.

MKadaner commented:
> P.P.S. Dogs vs cats?

Sorry, it's too personal a question.
